### ISSUE TITLE
pip --require-hashes ignores our == pinning in favor of >= from individual packages

### ISSUE THREAD

**Issue by erikrose** [certbot/certbot#5195]
On using certbot-auto, some people get this:

```
In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:     zope.interface>=4.1.0 from https://pypi.python.org/packages/21/a7/3cd033ea29c7614f3b41c593a16ae19b4470eb4eca0a4a883bec31d08955/zope.interface-4.4.3-cp27-cp27mu-manylinux1_x86_64.whl#md5=025a5f2dcafe75900e6725cc0e5a2288 (from zope.component==4.2.2->-r
/tmp/tmp.jhvCusu52p/letsencrypt-auto-requirements.txt (line 164))
```

We do, of course, pin zope.interface==4.1.3 in the requirements file, but pip somehow lets the specification in zope.component take precedence. ([pip doesn't have a real dependency resolver](https://github.com/pypa/pip/issues/988), so it's probably just reaching the zope.component one first.)

1. Try to reliably reproduce this.
2. Try to fix it by reordering the dependencies in our requirements files. Perhaps if we put zope.interface first, this will go away. (If it does, we should audit our dependencies for other such situations.)

Note that this is subtly different than #5163, where we legitimately don't have a dependency pinned. They might, however, have a common fix.

**Comment by bmw** [comment]
I unfortunately expect reproducing this will be hard, but the user who encountered this was running an updated version of Ubuntu 14.04 with certbot-auto 0.19.0.

**Comment by bmw** [comment]
As an aside, I'd argue that this is a bug in `pip`. If we can get a fix for this released upstream, we could update `pipstrap` again to upgrade people to a new enough version to avoid the problem.

**Comment by hal869** [comment]
anyone have an update on this, since last posts in 10/2017?

I was just informed that my issue (#6227), is _likely_ a dup of this ...
can anyone here take a look at #6227 to confirm? or not, &  to keep as separate issue?

**Comment by bmw** [comment]
Unfortunately not. [pip still doesn't have a dependency resolver](https://github.com/pypa/pip/issues/988) leaving us to try and figure out a way to fix this issue.

**Comment by bmw** [comment]
While https://github.com/pypa/pip/issues/988 remains unresolved and I couldn't quickly find any confirmation that this problem is blocked on that (in theory they could read the entire requirements file first before trying to resolve any dependencies), I haven't seen anyone hit this particular issue in a very long time so I'm closing this. If you're still hitting the problem, please comment or open a new issue.

**Comment by brainwane** [comment]
Anyone who is hitting this issue: please [try the beta of the new dependency resolver in pip 20.2](https://pip.pypa.io/en/latest/user_guide/#changes-to-the-pip-dependency-resolver-in-20-2-2020) and see whether that helps?

**Comment by erikrose** [comment]
I'm so excited that finally landed! Thanks, Sumana!
